generation/src/proof_generation/proofs/kore.py

```diff
@@ -108,11 +115,47 @@ def kore_exists(var: int) -> Notation:
     return Notation(
         'kore-exists',
         3,
-        _and(sorted_exists(var)(phi0, phi2), App(inhabitant_symbol, phi1)),
+        _and(sorted_exists(var)(phi0, phi2), kore_top(phi1)),
         f'( k∃ {var}:{0} . {2}):{1}',
     )
 
 
+@cache
+def kore_forall(var: int) -> Notation:
+    """kore_forall(inner_sort, outer_sort, pattern)"""
+    return Notation(
+        'kore-forall',
+        3,
+        _and(sorted_forall(var)(phi0, phi2), kore_top(phi1)),
+        f'( k∀ {var}:{0} . {2}):{1}',
+    )
+
+
+@cache
+def kore_functional(var: int) -> Notation:
+    """kore_functional(var)(sort_of_pattern, outer_sort, pattern)"""
+    return Notation(
+        'kore-functional',
+        3,
+        sorted_exists(var)(phi0, kore_equals(phi0, phi1, MetaVar(2, e_fresh=(EVar(var),)), EVar(var))),
+        'kore_is_functional({2}:{0}):{1}',
+    )
+
+
+@cache
+def kore_is_function(arg_sorts: list[Pattern]) -> Notation:
+    """kore_is_function(arity)(sort_of_symbol, outer_sort, symbol)"""
+    arg_len = len(arg_sorts)
+    notation_pattern: Pattern
+    notation_pattern = MetaVar(2, e_fresh=tuple([EVar(x) for x in range(arg_len + 1)]))
```

Comment:
```suggestion
    notation_pattern = MetaVar(2, e_fresh=tuple([EVar(x) for x in range(arg_len)]))
```

We want to have exactly `arg_len` of x's, which means this line as we start from 0 and go to `arg_len - 1`.

Then `y` will start at `arg_len` as you are using it in `kore_functional`

Reply:
What do you mean by `y`? EVars don't have string names, they just have integer id's. Only symbols have string names.
We need to account for the `arg_len` evars standing for the arguments as well as an extra one that is used to claim the functionality of the entire application.